### ISSUE TITLE
Try merging Resolver and Typechecker into one pass

### ISSUE THREAD

**Issue by kubouch** [nushell/new-nu-parser#32]
Merging the resolving and typechecking passes into one might help performance. The resulting pass can be called “semantic analysis”, or “Sema”.

**Comment by InnocentZero** [comment]
I'll take on this.

**Comment by kubouch** [comment]
OK, it would be good to know the performance tradeoff, since having the passes separate is preferred because it's more readable. The combined1000 benchmark is probably the most representative, but you can come up with some better one. You can use tango for A-B benchmarking via `path/to/benchmark1 compare path/to/benchmark2 ...args`. See also https://github.com/nushell/nushell/blob/4ff4e3f93de89702219d1559fd8cfa9122dd14b2/toolkit.nu#L500.

**Comment by kubouch** [comment]
@InnocentZero See https://github.com/nushell/new-nu-parser/issues/38, we should probably fix our benchmarks first before optimizing.

**Comment by kubouch** [comment]
The benchmarks are fixed, and the new parser seems to be at least 10x faster in most benchmarks, which is good. But the combinedX.nu benchmarks scale less efficiently than the old parser. Every 10x increase in code size, the new parser scales about 20x, while the old one is closer to 10x. So it would still be interesting to see how the performance would be with the two passes merged.

**Comment by sholderbach** [comment]
Let's analyze which parts don't scale well before rushing to make that single pass

**Comment by InnocentZero** [comment]
I guess we need finer benchmarking tests to see which parts don't scale well wrt code size. I can write a few tests if that's ok with all of you.

**Comment by kubouch** [comment]
Yeah, why not. We might also want to wait until the compiler is more complete, it's still missing several fundamental parts.
